File: backend/app/models/resource_metric.py
from datetime import datetime
from pydantic import Field
from beanie import Indexed, Link
from app.models.base import TimestampedDocument
from app.models.cloud_resource import CloudResource
from app.models.base import get_utc_now

class ResourceMetric(TimestampedDocument):
    resource: Link[CloudResource]
    cpu_utilization: float = Field(0.0, description="Percentage of CPU utilized")
    memory_utilization: float = Field(0.0, description="Percentage of memory utilized")
    storage_utilization: float = Field(0.0, description="Percentage of storage utilized")
    network_in: float = Field(0.0, description="Network incoming usage in MB")
    network_out: float = Field(0.0, description="Network outgoing usage in MB")
    disk_read: float = Field(0.0, description="Disk read operations/usage")
    disk_write: float = Field(0.0, description="Disk write operations/usage")
    uptime_hours: float = Field(0.0, description="Uptime in hours")
    instance_state: Indexed(str) = "running"
    metric_timestamp: Indexed(datetime) = Field(default_factory=get_utc_now)

    class Settings:
        name = "resource_metrics"
        indexes = [
            "metric_timestamp",
            "instance_state"
        ]

File: backend/src/models/ResourceMetric.ts
import mongoose, { Schema, Document } from 'mongoose';
import { BaseModelPlugin } from './BaseModelPlugin';
import crypto from 'crypto';
import { ICloudResource } from './CloudResource';

export interface IResourceMetric extends Document<any> {
    _id: Buffer | string;
    resource_id: string;
    cpu_utilization: number;
    memory_utilization: number;
    storage_utilization: number;
    network_in: number;
    network_out: number;
    disk_read: number;
    disk_write: number;
    uptime_hours: number;
    instance_state: string;
    metric_timestamp: Date;
    created_at: Date;
    updated_at: Date;
}

const resourceMetricSchema = new Schema<IResourceMetric>({
    _id: { type: 'UUID', default: () => crypto.randomUUID() },
    resource_id: { type: String, ref: 'CloudResource', required: true, index: true },
    cpu_utilization: { type: Number, default: 0.0 },
    memory_utilization: { type: Number, default: 0.0 },
    storage_utilization: { type: Number, default: 0.0 },
    network_in: { type: Number, default: 0.0 },
    network_out: { type: Number, default: 0.0 },
    disk_read: { type: Number, default: 0.0 },
    disk_write: { type: Number, default: 0.0 },
    uptime_hours: { type: Number, default: 0.0 },
    instance_state: { type: String, default: 'running', index: true },
    metric_timestamp: { type: Date, default: Date.now, index: true }
}, { collection: 'resource_metrics' });

resourceMetricSchema.plugin(BaseModelPlugin);

export const ResourceMetric = mongoose.model<IResourceMetric>('ResourceMetric', resourceMetricSchema);
